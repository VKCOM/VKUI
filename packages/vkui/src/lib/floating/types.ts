import type { Placement } from '@vkontakte/vkui-floating-ui/react-dom';

export type AutoPlacementType = 'auto' | 'auto-start' | 'auto-end';

export type PlacementWithAuto = AutoPlacementType | Placement;

export type {
  Placement,
  Middleware as UseFloatingMiddleware,
  UseFloatingData,
  Strategy as FloatingPositionStrategy,
} from '@vkontakte/vkui-floating-ui/react-dom';
