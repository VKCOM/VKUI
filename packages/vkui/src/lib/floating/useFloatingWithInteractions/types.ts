import type {
  Placement,
  UseFloatingData,
  UseFloatingMiddleware,
  UseFloatingRefs,
} from '../types/common';

export type InteractiveTriggerType = 'click' | 'hover' | 'focus';

export type ManualTriggerType = 'manual';

export type TriggerType = ManualTriggerType | InteractiveTriggerType | InteractiveTriggerType[];

export type ShownChangeReason =
  | 'click-outside'
  | 'escape-key'
  | 'click'
  | 'hover'
  | 'focus'
  | 'callback';

export type OnShownChange = (shown: boolean, reason?: ShownChangeReason) => void;

export interface UseFloatingWithInteractionsProps {
  placement?: Placement;
  middlewares?: UseFloatingMiddleware[];
  /**
   * Механика вызова всплывающего элемента.
   *
   * - `"click"` – показывается/скрывается только при нажатии.
   * - `"hover"` – будет показываться/скрывается при наведении/отведении мыши.
   * - `"focus"` – будет показываться/скрывается при фокусе/потере фокуса мыши.
   * - `"manual"` – будет показываться/скрывается только через свойство `shown`. `onShownChange`
   *    будет вызываться при нажатии за пределы целевого и всплывающего элементов, а также по кнопке
   *    ESC.
   *
   * > ⚠️`"hover"` на тач-устройствах будет работать как `"click"`, с одним лишь нюансом, что
   * > не будет закрываться при повторном нажатии на целевой элемент. Для закрытия необходимо нажать
   * > на область вне целевого элемента и выпадающего окна.
   */
  trigger?: TriggerType;
  /**
   * Количество миллисекунд, после которых произойдёт показ/скрытие всплывающего элемента
   * при наведении.
   *
   * > Чтобы задать разное время на показ и скрытие, передайте массив типа `[<показ>, <скрытие>]`.
   *
   * > Используется только для `trigger="hover"`.
   */
  hoverDelay?: number | [number, number];
  /**
   * При `trigger="hover"` закрывает всплывающий элемент при нажатии на целевой элемент.
   */
  closeAfterClick?: boolean;
  /**
   * Блокирует изменение состояния.
   */
  disabled?: boolean;
  /**
   * Отключает взаимодействие со всплывающим элементом.
   */
  disableInteractive?: boolean;
  /**
   * Отключает закрытие нажатием на область вне целевого и всплывающего элемента.
   */
  disableCloseOnClickOutside?: boolean;
  /**
   * Отключает закрытие нажатием на кнопку ESC.
   */
  disableCloseOnEscKey?: boolean;
  /**
   * Начальное состояние всплывающего элемента.
   */
  defaultShown?: boolean;
  /**
   * Если передан, то всплывающий элемент будет показано/скрыто в зависимости от значения свойства.
   */
  shown?: boolean;
  /**
   * Вызывается при каждом изменении видимости всплывающего элемента.
   */
  onShownChange?: OnShownChange;
}

export type ReferenceProps<T = HTMLElement> = Omit<
  React.HTMLAttributes<T>,
  keyof React.DOMAttributes<T>
> &
  Pick<React.DOMAttributes<T>, 'onMouseOver' | 'onMouseLeave' | 'onClick' | 'onFocus' | 'onBlur'>;

export type FloatingProps<T = HTMLElement> = Omit<
  React.HTMLAttributes<T>,
  keyof React.DOMAttributes<T> | 'style'
> & { style: React.CSSProperties } & Pick<
    React.DOMAttributes<T>,
    'onMouseOver' | 'onMouseLeave' | 'onClick' | 'onAnimationStart' | 'onAnimationEnd'
  >;

export interface UseFloatingWithInteractionsReturn<T extends HTMLElement = HTMLElement> {
  placement: Placement;
  shown: boolean;
  willBeHide: boolean;
  refs: UseFloatingRefs<T>;
  referenceProps: ReferenceProps<T>;
  floatingProps: FloatingProps<T>;
  middlewareData: UseFloatingData['middlewareData'];
  onClose(this: void): void;
  onEscapeKeyDown?(this: void): void;
  onRestoreFocus(this: void): boolean;
}
