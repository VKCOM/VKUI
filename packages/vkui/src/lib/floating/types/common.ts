import type {
  Placement,
  ReferenceType,
  UseFloatingReturn,
} from '@vkontakte/vkui-floating-ui/react-dom';

export type AutoPlacementType = 'auto' | 'auto-start' | 'auto-end';

export type PlacementWithAuto = AutoPlacementType | Placement;

export type {
  VirtualElement,
  UseFloatingOptions,
  ReferenceType,
  ArrowOptions,
  UseFloatingReturn,
  Placement,
  Middleware as UseFloatingMiddleware,
  UseFloatingData,
  Strategy as FloatingPositionStrategy,
} from '@vkontakte/vkui-floating-ui/react-dom';

export type UseFloatingRefs<RT extends ReferenceType = ReferenceType> =
  UseFloatingReturn<RT>['refs'];

export type OnPlacementChange = (nextPlacement: Placement) => void;
