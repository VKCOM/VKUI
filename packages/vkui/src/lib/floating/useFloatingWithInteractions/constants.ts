export const DEFAULT_TRIGGER = 'click';
