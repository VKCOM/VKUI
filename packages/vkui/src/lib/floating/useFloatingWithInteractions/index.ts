export { useFloatingWithInteractions } from './useFloatingWithInteractions';

export type {
  OnShownChange,
  UseFloatingWithInteractionsProps,
  UseFloatingWithInteractionsReturn,
} from './types';

export { DEFAULT_TRIGGER } from './constants';
