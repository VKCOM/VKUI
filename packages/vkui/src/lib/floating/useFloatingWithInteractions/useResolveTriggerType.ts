import type { TriggerType } from './types';

export const useResolveTriggerType = (triggerProp: TriggerType) =>
  (typeof triggerProp === 'string' ? [triggerProp] : triggerProp).reduce(
    (result, trigger) => {
      switch (trigger) {
        case 'click':
          result.triggerOnClick = true;
          return result;
        case 'hover':
          result.triggerOnHover = true;
          return result;
        case 'focus':
          result.triggerOnFocus = true;
          return result;
        case 'manual':
          return result;
      }
    },
    {
      triggerOnFocus: false,
      triggerOnClick: false,
      triggerOnHover: false,
    },
  );
