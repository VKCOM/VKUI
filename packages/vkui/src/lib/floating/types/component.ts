import * as React from 'react';
import type { UseFloatingMiddlewaresBootstrapOptions } from '../useFloatingMiddlewaresBootstrap';
import type {
  UseFloatingWithInteractionsProps,
  UseFloatingWithInteractionsReturn,
} from '../useFloatingWithInteractions';
import { OnPlacementChange } from './common';

/**
 * @private используйте алиасы, если для какого-то компонента нужно экспортировать тип
 */
export type FloatingContentRenderProp = (
  props: Pick<UseFloatingWithInteractionsReturn, 'onClose'>,
) => React.ReactNode;

/**
 * Общий API для всплывающих элементов.
 *
 * @private
 */
export interface FloatingComponentProps
  extends UseFloatingMiddlewaresBootstrapOptions,
    Omit<UseFloatingWithInteractionsProps, 'placement' | 'middlewares'> {
  /**
   * Содержимое всплывающего окна.
   *
   * При передаче контента в виде [render prop](https://react.dev/reference/react/cloneElement#passing-data-with-a-render-prop),
   * в аргументе функции можно получить метод `onClose`, с помощью которого можно программно закрывать
   * всплывающее окно.
   */
  content?: React.ReactNode | FloatingContentRenderProp;
  /**
   * Целевой элемент. Всплывающее окно появится возле него.
   *
   * > ⚠️ Если это пользовательский компонент, то он должен:
   * > 1. предоставлять параметры либо `getRootRef`, либо `ref` (cм. `React.forwardRef()`) для получения ссылки на DOM-узел;
   * > 2. принимать DOM атрибуты и события.
   */
  children?: React.ReactElement;
  /**
   * Нужно ли при навигации с клавиатуры авто-фокусироваться на всплывающий элемент.
   */
  autoFocus?: boolean;
  /**
   * Нужно ли после закрытия всплывающего элемента возвращать фокус на предыдущий активный элемент.
   */
  restoreFocus?: boolean;
  /**
   * Перебивает zIndex заданный по умолчанию.
   */
  zIndex?: number | string;
  /**
   * По умолчанию используется document.body.
   */
  usePortal?: boolean | HTMLElement | React.RefObject<HTMLElement>;
  /**
   * В зависимости от области видимости, позиция может смениться на более оптимальную,
   * чтобы всплывающий элемент вместился в эту область видимости.
   */
  onPlacementChange?: OnPlacementChange;
}
