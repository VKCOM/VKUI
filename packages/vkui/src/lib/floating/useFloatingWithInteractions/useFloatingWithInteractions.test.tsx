import * as React from 'react';
import { act, render, renderHook, waitFor } from '@testing-library/react';
import {
  AppRootContext,
  DEFAULT_APP_ROOT_CONTEXT_VALUE,
} from '../../../components/AppRoot/AppRootContext';
import { FocusTrap } from '../../../components/FocusTrap/FocusTrap';
import { fireEventPatch, userEvent } from '../../../testing/utils';
import { ShownChangeReason } from './types';
import { useFloatingWithInteractions } from './useFloatingWithInteractions';

const TestComponent = ({
  restoreFocus,
  hookResultRef,
  keyboardInput = false,
}: {
  restoreFocus?: boolean;
  hookResultRef: {
    current: ReturnType<typeof useFloatingWithInteractions<HTMLButtonElement>>;
  };
  keyboardInput?: boolean;
}) => {
  const { shown, refs, referenceProps, floatingProps } = hookResultRef.current;

  return (
    <AppRootContext.Provider value={{ ...DEFAULT_APP_ROOT_CONTEXT_VALUE, keyboardInput }}>
      <button ref={refs.reference} {...referenceProps}>
        Reference
      </button>
      {shown ? (
        <span ref={refs.floating} {...floatingProps}>
          <FocusTrap
            autoFocus
            restoreFocus={restoreFocus ? hookResultRef.current.onRestoreFocus : restoreFocus}
            onClose={hookResultRef.current.onEscapeKeyDown}
          >
            <input data-testid="focus-trap" type="text" defaultValue="Reference" />
          </FocusTrap>
        </span>
      ) : null}
    </AppRootContext.Provider>
  );
};

describe(useFloatingWithInteractions, () => {
  describe('test cases', () => {
    it.each([
      {
        trigger: 'focus' as const,
        openBy: 'focus' as const,
        closeBy: 'blur' as const, // 'escape-key' и 'click-outside' для focus вынесены в отдельные тесты
      },
      { trigger: 'click' as const, openBy: 'click' as const, closeBy: 'click' as const },
      { trigger: 'click' as const, openBy: 'click' as const, closeBy: 'escape-key' as const },
      { trigger: 'click' as const, openBy: 'click' as const, closeBy: 'click-outside' as const },
      { trigger: 'hover' as const, openBy: 'mouseOver' as const, closeBy: 'mouseLeave' as const },
      { trigger: 'hover' as const, openBy: 'mouseOver' as const, closeBy: 'escape-key' as const },
      { trigger: 'hover' as const, openBy: 'mouseOver' as const, closeBy: 'click-outside' as const }, // prettier-ignore
    ])(
      'should shown by $openBy event  and hidden by $closeBy event (trigger: $trigger)',
      async ({ trigger, openBy, closeBy }) => {
        const onShownChange = jest.fn();
        const { result } = renderHook(() =>
          useFloatingWithInteractions<HTMLButtonElement>({
            defaultShown: false,
            trigger: trigger,
            onShownChange,
          }),
        );
        const { rerender } = render(<TestComponent hookResultRef={result} />);
        await waitFor(() => {
          expect(result.current.shown).toBeFalsy();
          expect(onShownChange).toHaveBeenCalledTimes(0);
          expect(onShownChange).not.toHaveBeenCalled();
        });

        await fireEventPatch(result.current.refs.reference.current, openBy);
        rerender(<TestComponent hookResultRef={result} />);
        await waitFor(() => {
          expect(result.current.shown).toBeTruthy();
          expect(onShownChange).toHaveBeenCalledTimes(1);
          expect(onShownChange).toHaveBeenLastCalledWith(true, trigger);
        });

        let closeReason: ShownChangeReason = trigger;
        switch (closeBy) {
          case 'escape-key':
            closeReason = closeBy;
            jest.useFakeTimers();
            await userEvent.keyboard('{Escape}');
            act(() => {
              jest.runOnlyPendingTimers();
              jest.useRealTimers();
            });
            break;
          case 'click-outside':
            closeReason = closeBy;
            jest.useFakeTimers();
            await userEvent.click(document.body);
            act(() => {
              jest.runOnlyPendingTimers();
              jest.useRealTimers();
            });
            break;
          default:
            await fireEventPatch(result.current.refs.reference.current, closeBy);
        }
        rerender(<TestComponent hookResultRef={result} />);
        await waitFor(() => {
          expect(result.current.shown).toBeFalsy();
          expect(onShownChange).toHaveBeenCalledTimes(2);
          expect(onShownChange).toHaveBeenLastCalledWith(false, closeReason);
        });

        await fireEventPatch(result.current.refs.reference.current, openBy);
        rerender(<TestComponent hookResultRef={result} />);
        await waitFor(() => {
          expect(result.current.shown).toBeTruthy();
          expect(onShownChange).toHaveBeenCalledTimes(3);
          expect(onShownChange).toHaveBeenLastCalledWith(true, trigger);
        });
      },
    );

    it.each([
      { trigger: 'focus' as const, openBy: 'focus' as const, closeBy: 'escape-key' as const },
      { trigger: 'focus' as const, openBy: 'focus' as const, closeBy: 'click-outside' as const },
    ])(
      'should shown by $openBy event  and hidden by $closeBy event (trigger: $trigger)',
      async ({ trigger, openBy, closeBy }) => {
        const shouldClosedByClickOutside = closeBy === 'click-outside';
        const onShownChange = jest.fn();
        const { result } = renderHook(() =>
          useFloatingWithInteractions<HTMLButtonElement>({
            defaultShown: false,
            trigger: 'focus',
            onShownChange,
          }),
        );
        const { rerender } = render(<TestComponent hookResultRef={result} />);
        await waitFor(() => {
          expect(result.current.shown).toBeFalsy();
          expect(onShownChange).toHaveBeenCalledTimes(0);
          expect(onShownChange).not.toHaveBeenCalled();
        });

        await fireEventPatch(result.current.refs.reference.current, openBy);

        rerender(<TestComponent hookResultRef={result} />);
        await waitFor(() => {
          expect(result.current.shown).toBeTruthy();
          expect(onShownChange).toHaveBeenCalledTimes(1);
          expect(onShownChange).toHaveBeenLastCalledWith(true, trigger);
        });

        jest.useFakeTimers();
        if (shouldClosedByClickOutside) {
          await userEvent.click(document.body);
        } else {
          await userEvent.keyboard('{Escape}');
        }
        act(() => {
          jest.runOnlyPendingTimers();
          jest.useRealTimers();
        });

        rerender(<TestComponent hookResultRef={result} />);
        await waitFor(() => {
          if (shouldClosedByClickOutside) {
            expect(document.activeElement).toBe(result.current.refs.reference.current);
          }
          expect(result.current.shown).toBeFalsy();
          expect(onShownChange).toHaveBeenCalledTimes(2);
          expect(onShownChange).toHaveBeenLastCalledWith(false, closeBy);
        });

        if (shouldClosedByClickOutside) {
          expect(document.activeElement).toBe(result.current.refs.reference.current);
        }
        // Если всплывающее окно закрылось из-за 'escape-key' или 'click-outside', то при
        // возвращении фокуса оно не должно открыться повторно.
        expect(onShownChange).toHaveBeenCalledTimes(2);
        expect(result.current.shown).toBeFalsy();

        await fireEventPatch(result.current.refs.reference.current, 'focus');

        rerender(<TestComponent hookResultRef={result} />);
        await waitFor(() => {
          expect(onShownChange).toHaveBeenCalledTimes(2);
          expect(result.current.shown).toBeFalsy();
        });

        await fireEventPatch(result.current.refs.reference.current, 'blur'); // Сбрасываем блокировку фокуса

        rerender(<TestComponent hookResultRef={result} />);
        await waitFor(() => {
          expect(result.current.shown).toBeFalsy();
          expect(onShownChange).toHaveBeenCalledTimes(2);
          expect(onShownChange).toHaveBeenLastCalledWith(false, closeBy);
        });

        await fireEventPatch(result.current.refs.reference.current, openBy);

        rerender(<TestComponent hookResultRef={result} />);
        await waitFor(() => {
          expect(result.current.shown).toBeTruthy();
          expect(onShownChange).toHaveBeenCalledTimes(3);
          expect(onShownChange).toHaveBeenLastCalledWith(true, trigger);
        });
      },
    );

    it('should stayed shown if hover to floating element', async () => {
      const { result } = renderHook(() =>
        useFloatingWithInteractions<HTMLButtonElement>({
          defaultShown: false,
          hoverDelay: 0,
          trigger: 'hover',
        }),
      );
      const { rerender } = render(<TestComponent hookResultRef={result} />);
      await waitFor(() => expect(result.current.shown).toBeFalsy());

      await fireEventPatch(result.current.refs.reference.current, 'mouseOver');
      rerender(<TestComponent hookResultRef={result} />);
      await waitFor(() => expect(result.current.shown).toBeTruthy());

      await fireEventPatch(result.current.refs.floating.current, 'mouseOver');
      await waitFor(() => expect(result.current.shown).toBeTruthy());

      await fireEventPatch(result.current.refs.reference.current, 'mouseOver');
      await waitFor(() => expect(result.current.shown).toBeTruthy());

      await fireEventPatch(result.current.refs.floating.current, 'mouseOver');
      await waitFor(() => expect(result.current.shown).toBeTruthy());

      await fireEventPatch(result.current.refs.floating.current, 'mouseLeave');
      rerender(<TestComponent hookResultRef={result} />);
      await waitFor(() => expect(result.current.shown).toBeFalsy());
    });

    it.each([true, false])('should restore focus: %s', async (restoreFocus) => {
      const { result } = renderHook(() =>
        useFloatingWithInteractions<HTMLButtonElement>({
          defaultShown: false,
          trigger: 'focus',
        }),
      );
      const testComponentRender = render(
        <TestComponent hookResultRef={result} restoreFocus={restoreFocus} keyboardInput />,
      );
      await waitFor(() => expect(result.current.shown).toBeFalsy());

      await fireEventPatch(result.current.refs.reference.current, 'focus');
      testComponentRender.rerender(
        <TestComponent hookResultRef={result} restoreFocus={restoreFocus} keyboardInput />,
      );
      await waitFor(() => expect(result.current.shown).toBeTruthy());

      jest.useFakeTimers();
      await userEvent.keyboard('{Tab}');
      act(() => {
        jest.runOnlyPendingTimers();
        jest.useRealTimers();
      });
      await waitFor(() => expect(testComponentRender.getByTestId('focus-trap')).toHaveFocus());

      jest.useFakeTimers();
      await userEvent.keyboard('{Escape}');
      act(() => {
        jest.runOnlyPendingTimers();
        jest.useRealTimers();
      });
      testComponentRender.rerender(
        <TestComponent hookResultRef={result} restoreFocus={restoreFocus} keyboardInput />,
      );
      await waitFor(() => {
        expect(result.current.shown).toBeFalsy();
        if (restoreFocus) {
          expect(result.current.refs.reference.current).toHaveFocus();
        } else {
          expect(document.body).toHaveFocus();
        }
      });
    });

    it('should block focus', async () => {
      const { result } = renderHook(() =>
        useFloatingWithInteractions<HTMLButtonElement>({
          defaultShown: false,
          trigger: 'focus',
        }),
      );
      const testComponentRender = render(<TestComponent hookResultRef={result} />);
      await waitFor(() => expect(result.current.shown).toBeFalsy());

      jest.useFakeTimers();
      await userEvent.click(result.current.refs.reference.current!);
      act(() => {
        jest.runOnlyPendingTimers();
        jest.useRealTimers();
      });
      testComponentRender.rerender(<TestComponent hookResultRef={result} />);
      await waitFor(() => {
        expect(result.current.shown).toBeTruthy();
        expect(document.activeElement).toBe(result.current.refs.reference.current);
      });

      jest.useFakeTimers();
      await userEvent.click(document.body);
      act(() => {
        jest.runOnlyPendingTimers();
        jest.useRealTimers();
      });
      testComponentRender.rerender(<TestComponent hookResultRef={result} />);
      await waitFor(() => {
        expect(document.activeElement).toBe(document.body);
        expect(result.current.shown).toBeFalsy();
      });
    });

    it.each(['focus' as const, 'click' as const])(
      'should prefer shown state by %s trigger',
      async (eventName) => {
        const { result } = renderHook(() =>
          useFloatingWithInteractions<HTMLButtonElement>({
            defaultShown: false,
            trigger: ['focus', 'click', 'hover'],
          }),
        );
        const testComponentRender = render(<TestComponent hookResultRef={result} />);
        await waitFor(() => expect(result.current.shown).toBeFalsy());

        await fireEventPatch(result.current.refs.reference.current, eventName);
        testComponentRender.rerender(<TestComponent hookResultRef={result} />);
        await waitFor(() => expect(result.current.shown).toBeTruthy());

        await fireEventPatch(result.current.refs.reference.current, 'mouseLeave');
        testComponentRender.rerender(<TestComponent hookResultRef={result} />);
        await waitFor(() => expect(result.current.shown).toBeTruthy());

        await fireEventPatch(
          result.current.refs.reference.current,
          eventName === 'focus' ? 'blur' : eventName,
        );
        testComponentRender.rerender(<TestComponent hookResultRef={result} />);
        await waitFor(() => expect(result.current.shown).toBeFalsy());

        await fireEventPatch(result.current.refs.reference.current, eventName);
        testComponentRender.rerender(<TestComponent hookResultRef={result} />);
        await waitFor(() => expect(result.current.shown).toBeTruthy());

        await fireEventPatch(result.current.refs.floating.current, 'mouseOver');
        testComponentRender.rerender(<TestComponent hookResultRef={result} />);
        await waitFor(() => expect(result.current.shown).toBeTruthy());

        await fireEventPatch(result.current.refs.floating.current, 'mouseLeave');
        testComponentRender.rerender(<TestComponent hookResultRef={result} />);
        await waitFor(() => expect(result.current.shown).toBeTruthy());
      },
    );

    it('should use manual state', async () => {
      const { result, rerender } = renderHook(
        (props) => useFloatingWithInteractions<HTMLButtonElement>(props),
        {
          initialProps: { shown: false, trigger: 'manual' as const },
        },
      );
      render(<TestComponent hookResultRef={result} />);
      await waitFor(() => expect(result.current.shown).toBeFalsy());

      rerender({ shown: true, trigger: 'manual' });
      await waitFor(() => expect(result.current.shown).toBeTruthy());
    });

    it('should shown/hidden with animations', async () => {
      const { result } = renderHook(() =>
        useFloatingWithInteractions<HTMLButtonElement>({
          defaultShown: false,
          trigger: 'click',
        }),
      );
      render(<TestComponent hookResultRef={result} />);
      await waitFor(() => expect(result.current.shown).toBeFalsy());

      // show
      await fireEventPatch(result.current.refs.reference.current, 'click');
      await waitFor(() => {
        expect(result.current.willBeHide).toBeFalsy();
        expect(result.current.shown).toBeTruthy();
      });

      const onAnimationStart = jest.spyOn(result.current.floatingProps, 'onAnimationStart');
      render(<TestComponent hookResultRef={result} />);

      await fireEventPatch(result.current.refs.floating.current, 'animationStart');
      await waitFor(() => expect(onAnimationStart).toHaveBeenCalledTimes(1));

      // hide
      await fireEventPatch(result.current.refs.reference.current, 'click');
      await waitFor(() => {
        expect(result.current.willBeHide).toBeTruthy();
        expect(result.current.shown).toBeTruthy();
      });

      const onAnimationEnd = jest.spyOn(result.current.floatingProps, 'onAnimationEnd');
      render(<TestComponent hookResultRef={result} />);

      await fireEventPatch(result.current.refs.floating.current, 'animationEnd');
      await waitFor(() => {
        expect(onAnimationEnd).toHaveBeenCalledTimes(1);
        expect(result.current.willBeHide).toBeFalsy();
        expect(result.current.shown).toBeFalsy();
      });
    });

    it('should close using the onClose()', async () => {
      const onShownChange = jest.fn();
      const { result } = renderHook(() =>
        useFloatingWithInteractions({ defaultShown: true, onShownChange }),
      );
      await waitFor(() => expect(result.current.shown).toBeTruthy());
      act(() => {
        result.current.onClose();
      });
      await waitFor(() => {
        expect(result.current.shown).toBeFalsy();
        expect(onShownChange).toHaveBeenCalledTimes(1);
        expect(onShownChange).toHaveBeenCalledWith(false, 'callback');
      });
    });
  });

  describe('tests with snapshot', () => {
    it('should return default values', () => {
      const { result } = renderHook(() => useFloatingWithInteractions({}));
      expect(result.current).toMatchSnapshot();
    });

    it('should be hidden state', () => {
      const { result } = renderHook(() =>
        useFloatingWithInteractions({
          defaultShown: false,
          trigger: ['focus', 'click', 'hover'],
        }),
      );
      expect(result.current).toMatchSnapshot();
    });

    it('should be shown state', () => {
      const { result } = renderHook(() =>
        useFloatingWithInteractions({
          defaultShown: true,
          trigger: ['focus', 'click', 'hover'],
        }),
      );
      expect(result.current).toMatchSnapshot();
    });

    it('should be shown state with disabled interactive feature', () => {
      const { result } = renderHook(() =>
        useFloatingWithInteractions({
          defaultShown: true,
          trigger: ['focus', 'click', 'hover'],
          disableInteractive: true,
        }),
      );
      expect(result.current).toMatchSnapshot();
    });

    it('should be shown state with disabled close on esc key and click outside events', () => {
      const { result } = renderHook(() =>
        useFloatingWithInteractions({
          defaultShown: true,
          trigger: ['focus', 'click', 'hover'],
          disableCloseOnClickOutside: true,
          disableCloseOnEscKey: true,
        }),
      );
      expect(result.current).toMatchSnapshot();
    });
  });
});
