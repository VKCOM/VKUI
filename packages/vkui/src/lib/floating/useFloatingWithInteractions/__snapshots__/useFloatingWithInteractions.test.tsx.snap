// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFloatingWithInteractions tests with snapshot should be hidden state 1`] = `
{
  "floatingProps": {
    "onMouseLeave": [Function],
    "onMouseOver": [Function],
    "style": {},
  },
  "middlewareData": {},
  "onClose": [Function],
  "onEscapeKeyDown": undefined,
  "onRestoreFocus": [Function],
  "placement": "bottom",
  "referenceProps": {
    "onBlur": [Function],
    "onClick": [Function],
    "onFocus": [Function],
    "onMouseLeave": [Function],
    "onMouseOver": [Function],
  },
  "refs": {
    "floating": {
      "current": null,
    },
    "reference": {
      "current": null,
    },
    "setFloating": [Function],
    "setReference": [Function],
  },
  "shown": false,
  "willBeHide": false,
}
`;

exports[`useFloatingWithInteractions tests with snapshot should be shown state 1`] = `
{
  "floatingProps": {
    "onAnimationEnd": [Function],
    "onAnimationStart": [Function],
    "onMouseLeave": [Function],
    "onMouseOver": [Function],
    "style": {
      "bottom": "auto",
      "left": 0,
      "position": "fixed",
      "right": "auto",
      "top": 0,
      "width": "max-content",
    },
  },
  "middlewareData": {},
  "onClose": [Function],
  "onEscapeKeyDown": [Function],
  "onRestoreFocus": [Function],
  "placement": "bottom",
  "referenceProps": {
    "onBlur": [Function],
    "onClick": [Function],
    "onFocus": [Function],
    "onMouseLeave": [Function],
    "onMouseOver": [Function],
  },
  "refs": {
    "floating": {
      "current": null,
    },
    "reference": {
      "current": null,
    },
    "setFloating": [Function],
    "setReference": [Function],
  },
  "shown": true,
  "willBeHide": false,
}
`;

exports[`useFloatingWithInteractions tests with snapshot should be shown state with disabled close on esc key and click outside events 1`] = `
{
  "floatingProps": {
    "onAnimationEnd": [Function],
    "onAnimationStart": [Function],
    "onMouseLeave": [Function],
    "onMouseOver": [Function],
    "style": {
      "bottom": "auto",
      "left": 0,
      "position": "fixed",
      "right": "auto",
      "top": 0,
      "width": "max-content",
    },
  },
  "middlewareData": {},
  "onClose": [Function],
  "onEscapeKeyDown": undefined,
  "onRestoreFocus": [Function],
  "placement": "bottom",
  "referenceProps": {
    "onBlur": [Function],
    "onClick": [Function],
    "onFocus": [Function],
    "onMouseLeave": [Function],
    "onMouseOver": [Function],
  },
  "refs": {
    "floating": {
      "current": null,
    },
    "reference": {
      "current": null,
    },
    "setFloating": [Function],
    "setReference": [Function],
  },
  "shown": true,
  "willBeHide": false,
}
`;

exports[`useFloatingWithInteractions tests with snapshot should be shown state with disabled interactive feature 1`] = `
{
  "floatingProps": {
    "onAnimationEnd": [Function],
    "onAnimationStart": [Function],
    "style": {
      "bottom": "auto",
      "left": 0,
      "pointerEvents": "none",
      "position": "fixed",
      "right": "auto",
      "top": 0,
      "width": "max-content",
    },
  },
  "middlewareData": {},
  "onClose": [Function],
  "onEscapeKeyDown": [Function],
  "onRestoreFocus": [Function],
  "placement": "bottom",
  "referenceProps": {
    "onBlur": [Function],
    "onClick": [Function],
    "onFocus": [Function],
    "onMouseLeave": [Function],
    "onMouseOver": [Function],
  },
  "refs": {
    "floating": {
      "current": null,
    },
    "reference": {
      "current": null,
    },
    "setFloating": [Function],
    "setReference": [Function],
  },
  "shown": true,
  "willBeHide": false,
}
`;

exports[`useFloatingWithInteractions tests with snapshot should return default values 1`] = `
{
  "floatingProps": {
    "style": {},
  },
  "middlewareData": {},
  "onClose": [Function],
  "onEscapeKeyDown": undefined,
  "onRestoreFocus": [Function],
  "placement": "bottom",
  "referenceProps": {
    "onClick": [Function],
  },
  "refs": {
    "floating": {
      "current": null,
    },
    "reference": {
      "current": null,
    },
    "setFloating": [Function],
    "setReference": [Function],
  },
  "shown": false,
  "willBeHide": false,
}
`;
